Project Files/DLLSources/CvPlayer.cpp

```diff
@@ -20168,12 +20151,31 @@ void CvPlayer::doLbD()
 	//getting GameSpeedModifiert in percent
 	int train_percent = GC.getGameSpeedInfo(GC.getGameINLINE().getGameSpeedType()).getTrainPercent();
 
-	//moddifying Expert values with GameSpeed
-	chance_increase_expert = chance_increase_expert / train_percent / 100;
+	//moddifying values with GameSpeed
+	chance_increase_expert = chance_increase_expert * 100 / train_percent;
 	pre_rounds_expert = pre_rounds_expert * train_percent / 100;
 
-	chance_increase_free = chance_increase_free / train_percent / 100;
+	chance_increase_free = chance_increase_free * 100 /  train_percent ;
+	if(chance_increase_free <= 0) {
+		chance_increase_free = 1;
+	}
+	if (chance_increase_expert <= 0) {
+		chance_increase_expert = 1;
+	}
 	pre_rounds_free = pre_rounds_free * train_percent / 100;
+	for (int iTrait = 0; iTrait < GC.getNumTraitInfos(); ++iTrait)
+	{
+		TraitTypes eTrait = (TraitTypes) iTrait;
+		if (eTrait != NO_TRAIT)
+		{
+			if (hasTrait(eTrait))
+			{
+				pre_rounds_expert *= 100;
+				pre_rounds_expert /= GC.getTraitInfo(eTrait).getLearningByDoingModifier() + 100;

```

Comment:
same comments as the previous modifier applying code. Also it would likely be a good idea to cache the combined modifier in CvPlayer. This will boost performance and make it easier to make it work with CivEffects later.

Reply:
I'm not sure how do you want me to cache this modifier.... by having an algorithm like this at CvPlayer?
```
CvPlayer::getPreRoundsExpert() {
  if preRoundsExpert == -1 then
      .. //calculate pre Rounds expert 
}
```
?